src/metabase/models/util/spec_update.clj

```diff
@@ -0,0 +1,162 @@
+(ns metabase.models.util.spec-update
+  "Perform Creation/Update/Deletion with a spec."
+  (:require
+   [clojure.string :as str]
+   [metabase.util :as u]
+   [metabase.util.log :as log]
+   [metabase.util.malli :as mu]
+   [toucan2.core :as t2]))
+
+(defn- format-path
+  [path]
+  (str (str/join #" / " (map #(if (keyword? %) (name %) %) path)) ":"))
+
+;; TODO: add :id-column, currently it assumes all models use :id
+(def ^:private Spec
+  [:schema {:registry {::spec [:map {:closed true}
+                               [:model                         [:fn #(isa? % :metabase/model)]]
+                               ;; whether this nested model is a sequentials with respect to the parent model
+                               [:multi-row?   {:optional true} :boolean]
+                               ;; the foreign key column in the nested model with respect to the parent model
+                               [:fk-column    {:optional true} :keyword]
+                               ;; A list of columns that should be compared to determine if the row has changed
+                               [:compare-row  {:optional true} [:sequential :keyword]]
+                               [:nested-specs {:optional true} [:map-of :keyword [:ref ::spec]]]]}}
+   [:ref ::spec]])
+
+(defn- compare-row-fn
+  [spec]
+  (if-let [compare-row (:compare-row spec)]
+    (fn [row]
+      (select-keys row compare-row))
+    identity))
+
+(declare do-map-update!)
+(declare do-update!*)
+
+(defn- handle-nested-updates!
+  [existing-row new-row nested-specs path]
+  (doseq [[k spec] nested-specs]
+    (do-update!* (get existing-row k) (get new-row k) spec (conj path k))))
+
+(defn- with-parent-id
+  [row nested-specs parent-id]
+  (merge row
+         (into {}
+               (for [[k {:keys [fk-column multi-row?]}] nested-specs]
+                 [k (if fk-column
+                      (if multi-row?
+                        (map #(assoc % fk-column parent-id) (get row k))
+                        (assoc (get row k) fk-column parent-id))
+                      row)]))))
+
+(defn- sanitize-row-fn
+  [{:keys [nested-specs] :as spec}]
+  (comp (compare-row-fn spec) #(apply dissoc % (keys nested-specs))))
+
+(defn- do-sequential-updates!
+  [existing-rows new-rows {:keys [model nested-specs] :as spec} path]
+  (let [{:keys [to-update
+                to-create
+                to-delete
+                to-skip]
+         :as _updates}    (u/row-diff existing-rows new-rows :to-compare (compare-row-fn spec))
+        sanitize-row      (sanitize-row-fn spec)]
+    (when (seq to-create)
+      (if nested-specs
+        (do
+          (log/debugf "%s nested spec found, creating rows one by one" (format-path path))
```

Comment:
I agree with Chris, this should be `tracef`

Reply:
done